copilot-language-server: init at 1.273.0

arunoruto opened the pull request:
This is the first time I am adding an npm package to nixpkgs. Feel free to correct me, if I did something wrong!

I am aware that lock files are not desired in nixpkgs (#327064), but the copilot ls is still in a private repo by GitHub, so there is not way to contact them to provide a lockfile themselves. If they ever decide to provide a lock file in their tarball, I would delete the one from nixpkgs and use theirs!

npmjs: https://www.npmjs.com/package/@github/copilot-language-server

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

arunoruto commented:
> Could you use [buildNpmPackage](https://nixos.org/manual/nixpkgs/stable/#javascript-buildNpmPackage) instead of adding to the nodePackages? see: #229475

I am totally aware that building with `buildNpmPackage` would be better, but it seems like GitHub isn't publishing the source code of the package... The npm package was the only reference they gave in [their blog post](https://github.blog/changelog/2025-02-10-copilot-language-server-sdk-is-now-available/). But if they publish it sometime in the future, I could build it from source!

Or do you know where the source code could be found?

donovanglover commented:
You can use `npm view @github/copilot-language-server` to get the tarball that can be used with `fetchurl`. [Example](https://github.com/NixOS/nixpkgs/pull/317222/commits/43ad022aa68cc014bc5cc03a1addd610e708047c#diff-9ea708dc978de622e5b56cea8737211193b89da3255dd7f717bb7c3534f921a8R7-R16).

arunoruto committed:
copilot-language-server: init at 1.273.0

arunoruto commented:
> You can use `npm view @github/copilot-language-server` to get the tarball that can be used with `fetchurl`. [Example](https://github.com/NixOS/nixpkgs/pull/317222/commits/43ad022aa68cc014bc5cc03a1addd610e708047c#diff-9ea708dc978de622e5b56cea8737211193b89da3255dd7f717bb7c3534f921a8R7-R16).

Thanks for the tips! I just made a new commit with the wanted changes and it seems to work, but I haven't tested it extensively...
Is there a way to enable auto update with npm packages? Or is a custom update script needed, as it has been done in the example you gave me?

brianmcgillion commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 381276`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>copilot-language-server</li>
  </ul>
</details>

---
### `aarch64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>copilot-language-server</li>
  </ul>
</details>

ngn999 commented:
I tested the package on arm64-darwin, and it works as expected. Could you please add support for the macOS platform as well? This enhancement would help broaden the package’s compatibility for macOS users.

Thanks for your work on this!

arunoruto commented:
> @arunoruto I tested the package on arm64-darwin, and it works as expected. Could you please add support for the macOS platform as well? This enhancement would help broaden the package’s compatibility for macOS users.
> 
> Thanks for your work on this!

I just added darwin as supported platforms (both arm and x86). While linux arm isn't officially supported, I added it and maybe someone can test it?

@ngn999 could you make a `nixpkgs-review` with a Mac and post the results here?

wattmto commented:
> I just added darwin as supported platforms (both arm and x86). While linux arm isn't officially supported, I added it and maybe someone can test it?

`aarch64-linux` seems not supported because there are no linux arm binary in `native` directory.

https://www.npmjs.com/package/@github/copilot-language-server?activeTab=code

arunoruto commented:
> > I just added darwin as supported platforms (both arm and x86). While linux arm isn't officially supported, I added it and maybe someone can test it?
> 
> `aarch64-linux` seems not supported because there are no linux arm binary in `native` directory.
> 
> https://www.npmjs.com/package/@github/copilot-language-server?activeTab=code

You are totally right, I just removed it now :)
One can also try to run it via node for different architectures, according to the README:
`node ./node_modules/@github/copilot-language-server/dist/language-server.js --version`

ngn999 commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 381276`

---
### `aarch64-darwin`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>copilot-language-server</li>
  </ul>
</details>

wattmto commented:
> Is there a way to enable auto update with npm packages? Or is a custom update script needed, as it has been done in the example you gave me?

I looked up other packages and I think there are no ways to update npm packages with tarball source automatically other than custom script.

arunoruto commented:
> Just a thought. It might be worth asking the maintainer of [nix-update](https://github.com/Mic92/nix-update) if they would accept a PR to add this functionality to `nix-update`. The script is all boilerplate. The version number and url are in the main recipe, and the tarball can be stripped from the end. Then you would be able to just do the `passthru.updateScript = nix-update-script {} ;` semantics for all the tar'd node packages.

I had the same idea while making the `update.sh` script. Maybe add a parameter for the location of the lock file, since it can be provided with the tarball, but doesn't have to (like in this case). This would be cool for node, go, rust, and other projects, but would make things a bit more complicated.

wattmto commented on pkgs/by-name/co/copilot-language-server/package.nix:
```diff
@@ -0,0 +1,42 @@
+  meta = {
+    description = "Use GitHub Copilot with any editor or IDE via the Language Server Protocol";
+    homepage = "https://github.com/features/copilot";
+    license = lib.licenses.cc-by-40;
```
Are there any sources that points this license? I only could find `"license": "https://docs.github.com/en/site-policy/github-terms/github-terms-for-additional-products-and-features",` in `package.json`.

arunoruto replied:
Seems like I jumped too fast to conclusion based on the licencing for the [advisory database](https://docs.github.com/en/site-policy/github-terms/github-terms-for-additional-products-and-features#advisory-database). But I wasn't able to find something better. Maybe make a custom one in this case?

wattmto replied:
custom one or unfree is fine. I think `GitHub Copilot Product Specific Terms` is applied.

https://docs.github.com/en/site-policy/github-terms/github-terms-for-additional-products-and-features#github-copilot

arunoruto replied:
I have adjusted the license to a custom, non-free one.
I am open for suggestions on what else to modify on the license end :)

wattmto reviewed (commented):
otherwise LGTM

arunoruto commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 381276`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>copilot-language-server</li>
  </ul>
</details>

donovanglover commented:
Please change the PR title and commit message from `init` to `init at <version>`

GaetanLepage reviewed (approved):
LGTM

arunoruto commented:
If someone is eager track the progress of the PR: https://nixpk.gs/pr-tracker.html?pr=381276

arunoruto commented:
It's available now on `nixos-unstable` with the updated version `1.275.0`.
